### PR TITLE
[PyROOT][DF] `AsNumpy` automatically converts to numpy arrays

### PR DESCRIPTION
**Pull request by lobis**
# This Pull request:

## Changes or fixes:

Calling `AsNumpy` on a dataframe (python api) now read `RVec` entries and `std::vector` as `numpy.ndarray` of the corresponding type. The implementation should be practically free besides some type checking.

It's not inmediate where to draw the line between what to convert into a `numpy.ndarray`. An `RVec` is clear but I also chose to convert `std::vector`, but not things such as `std::array`. Maybe `std::vector` shouldn't be converted. I haven't updated all the tests related to the conversion of `std::vector` so they are currently failing (I would like to have official confirmation).

We could also return a 2D (or N-D) regular array in the case the data is regular. This can be checked also without much overhead and the transformation into a regular array can be done via `numpy.stack`.  From the offline discussions we decided that we would leave this final optional step to the user.

There are some consequences of these changes: for instance when dealing with string data, if we have a vector of strings of different length, it won't be converted into a numpy array. We could optionaly check for this and return a numpy array of objects. (there is a deprecation warning when you try to make an array from jagged data). In the case where all strings are the same length, the output of `np.asarray` is to convert them into multidimensional arrays of chars, which may not agree with what the user is expecting. Perhaps we could handle the `std::string` case separately but in this case what other cases should we handle separately?

## Checklist:

- [x] tested changes locally
- [x] updated the docs (if necessary)

This PR fixes https://github.com/root-project/root/issues/7685

### CODE REVIEW

**Comment by enirolf** [comment]
@phsft-bot build

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py [hunk 1]
```diff
@@ -277,19 +282,24 @@ def GetValue(self):
...
                 if hasattr(cpp_reference, "__array_interface__"):
-                    tmp = numpy.asarray(cpp_reference) # This adopts the memory of the C++ object.
+                    tmp = numpy.asarray(cpp_reference)  # This adopts the memory of the C++ object.
                     self._py_arrays[column] = ndarray(tmp, self._result_ptrs[column])
                 else:
                     tmp = numpy.empty(len(cpp_reference), dtype=object)
                     for i, x in enumerate(cpp_reference):
-                        tmp[i] = x # This creates only the wrapping of the objects and does not copy.
+                        if any(is_templated_instance(x, class_name) for class_name in
+                               ["cppyy.gbl.std.vector", "cppyy.gbl.ROOT.VecOps.RVec"]) and all_same_length(x):
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py:298]: `all_same_lenght` works only if the iterable passed to it is more-than-1-dim IIUC, i.e.

```
auto x = ROOT::RVecI{1,2,3};
auto y = std::vector<int>{1,2,3};
```
These will not work with that function, as `len(next(iter(iterable)))` will return a TypeError because the iterable will give a single integer in this case, of which we can't compute a `len`. Instead this function expects something like
```
auto x = ROOT::RVec<ROOT::RVecI>{{1,2},{3,4},{5,6}};
```
To work properly.

**Review by vepadulano — CHANGES_REQUESTED** [review]
@lobis Thank you so much for this contribution! Something is not yet clear to me in the logic, see the comment about the `all_same_lenght` function. As things are now, the reproducer at https://github.com/root-project/root/pull/14717 doesn't work. I had to remove the call to `all_same_length` to get the expected result

**Comment by siliataider** [comment]
Hi @lobis ,
Thanks again for your contribution! I’ve rebased the branch and resolved conflicts to bring it up to date. Let me know if you were planning to continue working on it, otherwise I'd like to wrap it up by addressing the pending feedback from Vincenzo and finishing the implementation, if that’s okay with you!

**Comment by lobis** [comment]
> Hi @lobis , Thanks again for your contribution! I’ve rebased the branch and resolved conflicts to bring it up to date. Let me know if you were planning to continue working on it, otherwise I'd like to wrap it up by addressing the pending feedback from Vincenzo and finishing the implementation, if that’s okay with you!

Feel free to wrap it up! Unfortunately I do now have the time to finish this right now

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py (hunk 1 above)
Comment (siliataider) [reply bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py]: I removed the call to `all_same_length` as it's anyway unnecessary: calling `numpy.asarray` will automatically convert regularly shaped nested sequences into multidimensional arrays.

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py [hunk 2]
```diff
@@ -96,14 +96,33 @@ def pypowarray(numpyvec, pow):
 
 Eventually, you probably would like to inspect the content of the RDataFrame or process the data further
 with Python libraries. For this purpose, we provide the `AsNumpy()` function, which returns the columns
-of your RDataFrame as a dictionary of NumPy arrays. See a simple example below or a full tutorial [here](df026__AsNumpyArrays_8py.html).
+of your RDataFrame as a dictionary of NumPy arrays. See a few simple examples below or a full tutorial [here](df026__AsNumpyArrays_8py.html).
 
+\anchor asnumpy_scalar_columns
+##### Scalar Columns
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py:102]:
```suggestion
##### Scalar columns
```

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py [hunk 3]
```diff
@@ -96,14 +96,33 @@ def pypowarray(numpyvec, pow):
...
-print(cols["x"], cols["y"]) # the values of the cols dictionary are NumPy arrays
+rdf = ROOT.RDataFrame(10).Define("int_col", "1").Define("float_col", "2.3")
+print(rdf.AsNumpy(["int_col", "float_col"]))
+# Output: {'int_col': array([...], dtype=int32), 'float_col': array([...], dtype=float64)}
+~~~
+
+Columns containing non-fundamental types (e.g., objects, strings) will result in NumPy arrays with `dtype=object`.
+
+##### Vector Columns
+If your column contains vectors of fundamental types (e.g., std::vector<int>), `AsNumpy()` produces a NumPy array with `dtype=object` where each element is a NumPy array representing the vector data for that row.
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py:113]:
```suggestion
##### Collection columns
If your column contains collections of fundamental types (e.g., std::vector<int>), `AsNumpy()` produces a NumPy array with `dtype=object` where each element is a NumPy array representing the collection for its corresponding entry in the column
```

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py [hunk 4]
```diff
@@ -96,14 +96,33 @@ def pypowarray(numpyvec, pow):
...
+This mechanism works for regularly shaped arrays of fundamental types, where all nested arrays have the same length.
+
+~~~{.py}
+rdf = rdf.Define("v_col", "std::vector<int>{{1, 2, 3}}")
+print(rdf.AsNumpy(["v_col", "int_col", "float_col"]))
+# Output: {'v_col': array([array([1, 2, 3], dtype=int32), ...], dtype=object), ...}
 ~~~
 
+For vectors of non-fundamental types, `AsNumpy()` will fallback on the [default behavior](\ref asnumpy_scalar_columns) and produce a NumPy array with `dtype=object`.
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py:122]: This is not clear, because in general for all collection columns the dtype of the output array is going to be `object`

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py [hunk 5]
```diff
@@ -96,14 +96,33 @@ def pypowarray(numpyvec, pow):
...
+rdf = ROOT.RDataFrame(10).Define("int_col", "1").Define("float_col", "2.3")
+print(rdf.AsNumpy(["int_col", "float_col"]))
+# Output: {'int_col': array([...], dtype=int32), 'float_col': array([...], dtype=float64)}
+~~~
+
+Columns containing non-fundamental types (e.g., objects, strings) will result in NumPy arrays with `dtype=object`.
+
+##### Vector Columns
+If your column contains vectors of fundamental types (e.g., std::vector<int>), `AsNumpy()` produces a NumPy array with `dtype=object` where each element is a NumPy array representing the vector data for that row.
+This mechanism works for regularly shaped arrays of fundamental types, where all nested arrays have the same length.
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py:114]:
```suggestion
If the collection at a certain entry contains values of fundamental types, or if it is a regularly shaped multi-dimensional array of fundamental type, then the numpy array representing the collection for that entry will have the `dtype` associated with the value type of the collection, for example:
```

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py (hunk 4 above)
Comment (vepadulano) [reply bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py:122]: I would rephrase it with something like
```suggestion
If the collection at a certain entry contains values of non-fundamental type, `AsNumpy()` will fallback on the [default behavior](\ref asnumpy_scalar_columns) and produce a NumPy array with `dtype==object` for that collection.
```

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py [hunk 6]
```diff
@@ -96,14 +96,33 @@ def pypowarray(numpyvec, pow):
...
+
+~~~{.py}
+rdf = rdf.Define("v_col", "std::vector<int>{{1, 2, 3}}")
+print(rdf.AsNumpy(["v_col", "int_col", "float_col"]))
+# Output: {'v_col': array([array([1, 2, 3], dtype=int32), ...], dtype=object), ...}
 ~~~
 
+For vectors of non-fundamental types, `AsNumpy()` will fallback on the [default behavior](\ref asnumpy_scalar_columns) and produce a NumPy array with `dtype=object`.
+
+For jagged arrays, refer to the section on [interoperability with AwkwardArray](\ref awkward_interop).
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdataframe.py:124]:
```suggestion
For more complex collection types in your entries, e.g. when every entry has a jagged array value, refer to the section on [interoperability with AwkwardArray](\ref awkward_interop).
```

**Review by vepadulano — CHANGES_REQUESTED** [review]

File: bindings/pyroot/pythonizations/test/rdataframe_asnumpy.py [hunk 7]
```diff
@@ -264,7 +267,7 @@ def test_empty_array(self):
         df = ROOT.ROOT.RDataFrame(1).Define("x", "std::vector<float>()")
         npy = df.AsNumpy(["x"])
         self.assertEqual(npy["x"].size, 1)
-        self.assertTrue(npy["x"][0].empty())
```
[the remark this answers is not in the document]
Comment (siliataider) [reply bindings/pyroot/pythonizations/test/rdataframe_asnumpy.py:267]: With the change to how vector columns are now handled `npy["x"][0]` is a `numpy.ndarray` object which does not have the attribute `empty` so I replaced the check with `self.assertEqual(npy["x"][0].size, 0)`

**Review by vepadulano — APPROVED** [review]
LGTM, but let's ensure the CI is green before merging. I left a comment to avoid the failure on windows.
